### PR TITLE
Feature/validator fields match

### PR DESCRIPTION
**Pull request by LissetteIbnz**

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -1,14 +1,19 @@
 # lc-validator-fields-match
 
 This is a [lc-form-validation](https://github.com/Lemoncode/lcFormValidation) add-on that brings validation capabilities for:
-// TODO:
+* Validate if two fields match.
 
 Please, refer to [lc-form-validation](https://github.com/Lemoncode/lcFormValidation) to know more.
```
Comment (brauliodiez) [inline readme.md:7]: Let's complete this part (we will do the same in the DNI), keep the parapgrah indicating that is lc-form-validation...

First we have to add a more comprehensive explanation, for instance:

```
Validate if two fields match: check if two fields have the same value, sample usage: validate if password and repeat password fields match, validate if email and repeat email fields match...
```

We should add a link to codesandbox or stackblitz with sample running (first we have to publish it on github)

Add some code indicating a sample usage

File: package.json [hunk 2]
```diff
@@ -5,7 +5,7 @@
   "main": "index.js",
   "keywords": [],
   "author": "",
-  "license": "ISC",
+  "license": "MIT",
   "scripts": {
     "clean": "rimraf build",
```
Comment (brauliodiez) [inline package.json:10]: IMHO version shouldn't be 1.0.0 I would make it 0.1.0 we need to massage it a bit maybe (new versions will come)

File: src/fields-match.ts [hunk 3]
```diff
@@ -1,21 +1,25 @@
...
+
 export interface Fields {
   first: string;
   second: string;
 }
 
-export const validateFieldsMatch: FieldValidationFunction<Fields> = ({ first, second }) => {
+export const VALIDATION_TYPE = 'FIELDS-MATCH';
+
+export const validateFieldsMatch: FieldValidationFunction = ({ first, second }) => {
```
Comment (brauliodiez) [inline src/fields-match.ts]: I guess this won't work

FieldValidationFunction has three parameters:

- *value* that has changed
- *vm* viewmodel (all the data from the form, entity)
- *custom params*

It should be something lik

export const validateFieldMatch : FieldValidationFunction(value, vm, customParams);

In custom params you should pass the id of the fields (entity) that need to be compared, in this case we only need the id the of the "repeat email" or "repeat password", becuase the original value we got it already in our value param.

something like

MyViewModel {
  password: string;
  repeatpassword : string;
}

We want repeat password to hold the fields match validation

The call should be something like (let's say we have a viewModel, an input and the fieldId is 'password')

validateField(vm, e.target.value, 'password');

Something we have to consider (maybe is time to open lc-form-validation and add an update), what would happen if we have something like?

```javascript
MyViewModel {
 mysubObject: { 
  password: string;
  },
  repeatpassword : string;
}
``` 
To make this call work

```javascript
validateField(vm, e.target.value, 'mysubOject.password');
```

We have to use _lodash_ _getOr_, we have already getOr in lc-form-validation we have to consider wether to add a peer dependency, or expose it in lc-form, something to discuss later on (right now let's cover the basics).

File: package.json (hunk 2 above)
Comment (LissetteIbnz) [reply package.json:10]: Done.

File: src/fields-match.ts (hunk 3 above)
Comment (LissetteIbnz) [reply src/fields-match.ts]: @brauliodiez check this when you can, please

File: index.d.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+/**
+ * TODO:
+ * Validates an input value as a spanish National Identity Number (DNI)
+ * @param value Input value to be interpreted as DNI
+ */
```
Comment (brauliodiez) [inline index.d.ts:6]: Clean up, this comment is related to DNI

File: src/fields-match.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+export const VALIDATION_TYPE = 'FIELDS_MATCH';
+
+export const validateFieldsMatch = (value: string, vm: any, key: string): FieldValidationResult => {
+  const fieldValidationResult: FieldValidationResult = new FieldValidationResult();
+  let fieldsAreEqual = false;
+
+  const nestedKeyResult = getNestedKey(vm, key);
+  if (nestedKeyResult === undefined) {
+    fieldValidationResult.errorMessage = 'The field pass by customParams is wrong';
```
Comment (brauliodiez) [inline src/fields-match.ts]: Here I would just skip the validation (no error, for the user is not validation), maybe add a console.log warning message indicadting something like:

```typescript
console.warn(`FieldMatch Validator: field id ${key} to compare not found`)
```

File: src/fields-match.test.ts [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+      // Arrange
+      const viewModel = { password: 'correct', confirmPassword: 'correct' };
+      // Act
+      const result = validateFieldsMatch(viewModel.confirmPassword, viewModel, 'password');
+      // Assert
+      expect(result.succeeded).toBeTruthy();
+    });
+
+    describe('nested object =>', () => {
+      it('should invalidate when nested customParam is no belong to viewmodel', () => {
```
Comment (brauliodiez) [inline src/fields-match.test.ts]: does not belong

Comment (LissetteIbnz) [reply src/fields-match.test.ts]: 🤦‍♀️

File: src/fields-match.ts (hunk 5 above)
Comment (LissetteIbnz) [reply src/fields-match.ts]: Ok. Then I will update the test file with that.

File: readme.md (hunk 1 above)
Comment (LissetteIbnz) [reply readme.md:7]: When @Nasdan publishes it, I'll add the samples.

**Pull request merged**